npeps/npep-126-egress-traffic-control.md

```diff
@@ -76,13 +76,184 @@ selected cluster workloads to k8s-apiservers for securing the server.
+          node-role.kubernetes.io/control-plane:
+    ports:
+      - portNumber:
+          protocol: TCP
+          port: 443
+      - portNumber:
+          protocol: TCP
+          port: 6443
+```
+
+### Implementing egress traffic control towards external destinations
+
+This NPEP proposes to add a new type of `AdminNetworkPolicyEgressPeer` called `Networks`
+to be able to explicitly select external destination CIDRs outside the cluster.
+This peer type will not be supported in `AdminNetworkPolicyIngressPeer`.
+
+```
+// AdminNetworkPolicyEgressPeer defines a peer to allow traffic to.
+// Exactly one of the selector pointers must be set for a given peer. If a
+// consumer observes none of its fields are set, they must assume an unknown
+// option has been specified and fail closed.
+// +kubebuilder:validation:MaxProperties=1
+// +kubebuilder:validation:MinProperties=1
+type AdminNetworkPolicyEgressPeer struct {
+    <snipped>
+	// Networks defines a way to select IPs and CIDR blocks that represent
+  // entities that live outside the cluster as a peer.
+	// It is the list of NetworkCIDR (both v4 & v6) that can be used to define
+	// external destinations.
+  // You could use this to select internal cluster networks like podCIDR blocks, but it is
+  // recommended to use namespaces and pods peers instead in those cases.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=100
+	Networks []string `json:"networks,omitempty" validate:"omitempty,dive,cidr"`
+}
+```
+
+Note that the API expects `networks` to be a set of "external" destinations.
+However if user puts a podCIDR, nodeCIDR, serviceCIDR or other intra-cluster
+networks, it will work but it is recommended to use namespaces, pods, nodes
+peers to express such entities. Not all implementations can correctly define
+the boundary between "internal" and "external" destinations with respect to a
+Kubernetes cluster.
+
+Let's define ANP and BANP that refer to these external networks:
+```
+apiVersion: policy.networking.k8s.io/v1alpha1
+kind: AdminNetworkPolicy
+metadata:
+  name: node-as-egress-peers
+spec:
+  priority: 70
+  subject:
+    namespaces: {}
+  egress:
+  - name: "allow-all-egress-to-intranet"
+    action: "Allow"
+    to:
+    - networks:
+      - 192.0.2.0/24
+      - 203.0.113.0/24
+      - 198.51.100.0/24
+  - name: "allow-egress-to-external-dns"
+    action: "Allow"
+    to:
+    - networks:
+      - 194.0.2.0/24
+      - 205.0.113.15/32
+      - 199.51.100.10/32
+    ports:
+      - portNumber:
+          protocol: UDP
+          port: 53
+  - name: "pass-egress-to-internet"
+    action: "Pass"
+    to:
+    - networks:
+      - 0.0.0.0/0
+---
+apiVersion: policy.networking.k8s.io/v1alpha1
+kind: BaselineAdminNetworkPolicy
+metadata:
+  name: default
+spec:
+  subject:
+    namespaces: {}
+  egress:
+  - name: "deny-egress-to-internet"
+    action: "Deny"
+    to:
+    - networks:
+      - 0.0.0.0/0
```

Comment:
This denies all IPv4 traffic, not just egress to the internet, so you either need to rename it, or add another rule before it allowing pod-to-pod traffic.

Reply:
done.